### PR TITLE
kubemark: enable CRI for the hollow nodes

### LINKED ISSUE #41488: Enable CRI in hollow nodes for kubemark
CRI was disabled temporarily in hollow nodes by #41460 because hollow nodes were failing to create pods.

We need to fix this and enable CRI for hollow nodes.
```
E0215 06:11:42.010590       9 docker_sandbox.go:348] Failed to list checkpoints: failed to list checkpoint store: open /var/lib/dockershim/sandbox: no such file or directory
E0215 06:11:43.011674       9 docker_sandbox.go:348] Failed to list checkpoints: failed to list checkpoint store: open /var/lib/dockershim/sandbox: no such file or directory
E0215 06:11:44.012421       9 docker_sandbox.go:348] Failed to list checkpoints: failed to list checkpoint store: open /var/lib/dockershim/sandbox: no such file or directory
E0215 06:11:45.013715       9 docker_sandbox.go:348] Failed to list checkpoints: failed to list checkpoint store: open /var/lib/dockershim/sandbox: no such file or directory
E0215 06:11:46.014773       9 docker_sandbox.go:348] Failed to list checkpoints: failed to list checkpoint store: open /var/lib/dockershim/sandbox: no such file or directory
E0215 06:11:47.015716       9 docker_sandbox.go:348] Failed to list checkpoints: failed to list checkpoint store: open /var/lib/dockershim/sandbox: no such file or directory
E0215 06:11:48.016755       9 docker_sandbox.go:348] Failed to list checkpoints: failed to list checkpoint store: open /var/lib/dockershim/sandbox: no such file or directory
E0215 06:11:49.017643       9 docker_sandbox.go:348] Failed to list checkpoints: failed to list checkpoint store: open /var/lib/dockershim/sandbox: no such file or directory
E0215 06:11:50.018547       9 docker_sandbox.go:348] Failed to list checkpoints: failed to list checkpoint store: open /var/lib/dockershim/sandbox: no such file or directory
E0215 06:11:50.867382       9 eviction_manager.go:207] eviction manager: unexpected err: Missing stats for root container
E0215 06:11:51.019278       9 docker_sandbox.go:348] Failed to list checkpoints: failed to list checkpoint store: open /var/lib/dockershim/sandbox: no such file or directory
E0215 06:11:51.032207       9 kubelet.go:1616] Failed to check if disk space is available for the runtime: could not determine capacity for "runtime" fs. Info: {Usage:0 Capacity:0 Available:0 Timestamp:2017-02-15 06:11:51.032189255 +0000 UTC}
E0215 06:11:51.032236       9 kubelet.go:1624] Failed to check if disk space is available on the root partition: could not determine capacity for "root" fs. Info: {Usage:0 Capacity:0 Available:0 Timestamp:2017-02-15 06:11:51.032228044 +0000 UTC}
E0215 06:11:52.020037       9 docker_sandbox.go:348] Failed to list checkpoints: failed to list checkpoint store: open /var/lib/dockershim/sandbox: no such file or directory
E0215 06:11:53.020861       9 docker_sandbox.go:348] Failed to list checkpoints: failed to list checkpoint store: open /var/lib/dockershim/sandbox: no such file or directory
E0215 06:11:54.021635       9 docker_sandbox.go:348] Failed to list checkpoints: failed to list checkpoint store: open /var/lib/dockershim/sandbox: no such file or directory
E0215 06:11:55.022443       9 docker_sandbox.go:348] Failed to list checkpoints: failed to list checkpoint store: open /var/lib/dockershim/sandbox: no such file or directory
E0215 06:11:56.023353       9 docker_sandbox.go:348] Failed to list checkpoints: failed to list checkpoint store: open /var/lib/dockershim/sandbox: no such file or directory
E0215 06:11:56.996183       9 kubelet.go:1616] Failed to check if disk space is available for the runtime: could not determine capacity for "runtime" fs. Info: {Usage:0 Capacity:0 Available:0 Timestamp:2017-02-15 06:11:56.99616195 +0000 UTC}
E0215 06:11:56.996251       9 kubelet.go:1624] Failed to check if disk space is available on the root partition: could not determine capacity for "root" fs. Info: {Usage:0 Capacity:0 Available:0 Timestamp:2017-02-15 06:11:56.996234018 +0000 UTC}
E0215 06:11:57.024046       9 docker_sandbox.go:348] Failed to list checkpoints: failed to list checkpoint store: open /var/lib/dockershim/sandbox: no such file or directory
E0215 06:11:57.085429       9 kubelet.go:1616] Failed to check if disk space is available for the runtime: could not determine capacity for "runtime" fs. Info: {Usage:0 Capacity:0 Available:0 Timestamp:2017-02-15 06:11:56.99616195 +0000 UTC}
E0215 06:11:57.085454       9 kubelet.go:1624] Failed to check if disk space is available on the root partition: could not determine capacity for "root" fs. Info: {Usage:0 Capacity:0 Available:0 Timestamp:2017-02-15 06:11:56.996234018 +0000 UTC}
E0215 06:11:57.369977       9 remote_runtime.go:86] RunPodSandbox from runtime service failed: rpc error: code = 2 desc = checkpoint key "/k8s_POD_density15000-0-a2a37a33-f345-11e6-b17e-0242ac110007-m0s9r_e2e-tests-density-30-1-8qpl0_a708b4b9-f345-11e6-9e07-42010af00009_0" is not valid.
E0215 06:11:57.370105       9 kuberuntime_sandbox.go:54] CreatePodSandbox for pod "density15000-0-a2a37a33-f345-11e6-b17e-0242ac110007-m0s9r_e2e-tests-density-30-1-8qpl0(a708b4b9-f345-11e6-9e07-42010af00009)" failed: rpc error: code = 2 desc = checkpoint key "/k8s_POD_density15000-0-a2a37a33-f345-11e6-b17e-0242ac110007-m0s9r_e2e-tests-density-30-1-8qpl0_a708b4b9-f345-11e6-9e07-42010af00009_0" is not valid.
E0215 06:11:57.370132       9 kuberuntime_manager.go:620] createPodSandbox for pod "density15000-0-a2a37a33-f345-11e6-b17e-0242ac110007-m0s9r_e2e-tests-density-30-1-8qpl0(a708b4b9-f345-11e6-9e07-42010af00009)" failed: rpc error: code = 2 desc = checkpoint key "/k8s_POD_density15000-0-a2a37a33-f345-11e6-b17e-0242ac110007-m0s9r_e2e-tests-density-30-1-8qpl0_a708b4b9-f345-11e6-9e07-42010af00009_0" is not valid.
E0215 06:11:57.370260       9 pod_workers.go:182] Error syncing pod a708b4b9-f345-11e6-9e07-42010af00009 ("density15000-0-a2a37a33-f345-11e6-b17e-0242ac110007-m0s9r_e2e-tests-density-30-1-8qpl0(a708b4b9-f345-11e6-9e07-42010af00009)"), skipping: failed to "CreatePodSandbox" for "density15000-0-a2a37a33-f345-11e6-b17e-0242ac110007-m0s9r_e2e-tests-density-30-1-8qpl0(a708b4b9-f345-11e6-9e07-42010af00009)" with CreatePodSandboxError: "CreatePodSandbox for pod \"density15000-0-a2a37a33-f345-11e6-b17e-0242ac110007-m0s9r_e2e-tests-density-30-1-8qpl0(a708b4b9-f345-11e6-9e07-42010af00009)\" failed: rpc error: code = 2 desc = checkpoint key \"/k8s_POD_density15000-0-a2a37a33-f345-11e6-b17e-0242ac110007-m0s9r_e2e-tests-density-30-1-8qpl0_a708b4b9-f345-11e6-9e07-42010af00009_0\" is not valid."
```

@Random-Liu did a quick triage and found that the container ids used by the fake docker client are not valid checkpoint key names:https://github.com/kubernetes/kubernetes/pull/41460#issuecomment-279950446

/cc @freehan @Random-Liu @wojtek-t

### PR DESCRIPTION
**Pull request by yujuhong**
This fixes #41488

### CODE REVIEW

**Comment by yujuhong** [comment]
I'll watch the kubemark-100/500 after this is merged.

**Comment by yujuhong** [comment]
```
I0228 09:38:56.836] godep v79 (linux/amd64/go1.7.4)
W0228 09:41:20.696] godep: error downloading dep (vbom.ml/util/sortorder): unrecognized import path "vbom.ml/util/sortorder"
W0228 09:41:20.697] godep: Error downloading some deps. Aborting restore and check.
W0228 09:41:20.698] + godep restore
W0228 09:41:31.683] + ./hack/install-etcd.sh
W0228 09:43:39.161] curl: (7) Failed to connect to github-cloud.s3.amazonaws.com port 443: Connection timed out
I0228 09:43:39.262] Downloading https://github.com/coreos/etcd/releases/download/v3.0.17/etcd-v3.0.17-linux-amd64.tar.gz failed. 4 retries left.
W0228 09:45:47.674] curl: (7) Failed to connect to github-cloud.s3.amazonaws.com port 443: Connection timed out
I0228 09:45:47.775] Downloading https://github.com/coreos/etcd/releases/download/v3.0.17/etcd-v3.0.17-linux-amd64.tar.gz failed. 3 retries left.
W0228 09:47:56.057] curl: (7) Failed to connect to github-cloud.s3.amazonaws.com port 443: Connection timed out
I0228 09:47:56.159] Downloading https://github.com/coreos/etcd/releases/download/v3.0.17/etcd-v3.0.17-linux-amd64.tar.gz failed. 2 retries left.
W0228 09:50:04.442] curl: (7) Failed to connect to github-cloud.s3.amazonaws.com port 443: Connection timed out
I0228 09:50:04.543] Downloading https://github.com/coreos/etcd/releases/download/v3.0.17/etcd-v3.0.17-linux-amd64.tar.gz failed. 1 retries left.
W0228 09:52:12.954] curl: (7) Failed to connect to github-cloud.s3.amazonaws.com port 443: Connection timed out
I0228 09:52:13.055] Downloading https://github.com/coreos/etcd/releases/download/v3.0.17/etcd-v3.0.17-linux-amd64.tar.gz failed. 0 retries left.
W0228 09:52:13.981] !!! [0228 09:52:13] Call tree:
```
Failure unrelated to this PR.

**Comment by Random-Liu** [comment]
I believe fixes in fake docker client were also reverted in https://github.com/kubernetes/kubernetes/pull/41693. We should get them in, too.

**Comment by yujuhong** [comment]
@Random-Liu, I didn't realize that was reverted too. Added the commit.

**Comment by yujuhong** [comment]
@k8s-bot kops aws e2e test this

**Comment by yujuhong** [comment]
@wojtek-t, need another review since I included @Random-Liu's commit.

**Comment by yujuhong** [comment]
Rebased. Reapplying the lgtm.

**Comment by Random-Liu** [comment]
Still seems to be something wrong
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/empty.go:52
Expected error:
    <*errors.errorString | 0xc4204be760>: {
        s: "Only 0 pods started out of 1",
    }
    Only 0 pods started out of 1
not to have occurred
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/empty.go:51
```

Still no hollow node logs. :p

**Comment by yujuhong** [comment]
> Still no hollow node logs. :p

:(

**Comment by yujuhong** [comment]
> Still no hollow node logs. :p

@shyamjvs is this expected?

**Comment by shyamjvs** [comment]
@yujuhong It's definitely not expected. And if there indeed was a log file for hollow-kubelet/kubeproxy (which would've been if the pod started), then it would've got copied. So it seems like it never really reached the stage of running the container command. Something failed prior to that. Can you reproduce this locally @Random-Liu by running small cluster?

**Comment by yujuhong** [comment]
@shyamjvs the hollow nodes were confirmed running and ready before the test tried to start the test pod. Whether the pod is running or not, the hollow nodes should have produced logs.

**Comment by shyamjvs** [comment]
That's interesting. Now that I just checked, it seems like the logs are missing even for successful runs. Let me take a look into it.

**Comment by shyamjvs** [comment]
@yujuhong Oops, seems like we never turned on `ENABLE_HOLLOW_NODE_LOGGING` for the PR pre-submit test. I mistook it for the kubemark-5-gce test previously, sorry. I'll enable this in a new PR. Thanks for pointing:)

**Comment by yujuhong** [comment]
> @yujuhong Oops, seems like we never turned on ENABLE_HOLLOW_NODE_LOGGING for the PR pre-submit test. I mistook it for the kubemark-5-gce test previously, sorry. I'll enable this in a new PR. Thanks for pointing:)

Thanks! It'd help us debug more efficiently :-)

**Comment by yujuhong** [comment]
@k8s-bot kubemark e2e test this

**Comment by calebamiles** [comment]
@k8s-bot gci gce e2e test this

**Comment by yujuhong** [comment]
Patching the fake docker client to correctly returns errors when an image does not exist. Added @Random-Liu to review the new commit.

**Comment by yujuhong** [comment]
@k8s-bot cross build this. issue: #42443
@k8s-bot kops aws e2e test this

**Comment by yujuhong** [comment]
Ready for review.

**Comment by yujuhong** [comment]
The tests failures are irrelevant and will be fixed by other PRs.

**Comment by Random-Liu** [comment]
@yujuhong Will review soon. :)

**Comment by yujuhong** [comment]
> @yujuhong Will review soon. :)

Thanks! Hopefully this can get in before it's broken again...

**Comment by yujuhong** [comment]
@k8s-bot kops aws e2e test this

**Comment by yujuhong** [comment]
@k8s-bot kops aws e2e test this

**Comment by yujuhong** [comment]
@Random-Liu ping!

File: pkg/kubelet/dockertools/fake_docker_client.go [hunk 1]
```diff
@@ -313,39 +315,40 @@ func (f *FakeDockerClient) AssertCreatedByName(created []string) error {
...
-	sort.StringSlice(actualStopped).Sort()
-	if !reflect.DeepEqual(stopped, actualStopped) {
-		return fmt.Errorf("expected %#v, got %#v", stopped, actualStopped)
+	return sortedStringSlicesEqual(stopped, actualStopped)
+}
+
+func (f *FakeDockerClient) AssertImagesPulled(pulled []string) error {
+	f.Lock()
+	defer f.Unlock()
+	// Copy stopped to avoid modifying it.
```
Comment (Random-Liu) [inline pkg/kubelet/dockertools/fake_docker_client.go]: s/stopped/pulled.

File: pkg/kubelet/dockertools/fake_docker_client.go [hunk 2]
```diff
@@ -454,7 +457,14 @@ func (f *FakeDockerClient) InspectImageByRef(name string) (*dockertypes.ImageIns
 	defer f.Unlock()
 	f.appendCalled(calledDetail{name: "inspect_image"})
 	err := f.popError("inspect_image")
-	return f.Image, err
+	if result, ok := f.ImageInspects[name]; ok {
```
Comment (Random-Liu) [inline pkg/kubelet/dockertools/fake_docker_client.go]: nit: The logic here is a little bit anti-intuitive.

Will it break any existing test if we change this to:
```
if err != nil {
  return nil, err
}
result, ok := f.ImageInspects[name]
if !ok {
  return nil, ImageNotFoundError{name}
}
return result, nil
}
```

File: pkg/kubelet/dockertools/fake_docker_client.go [hunk 3]
```diff
@@ -464,7 +474,14 @@ func (f *FakeDockerClient) InspectImageByID(name string) (*dockertypes.ImageInsp
...
```
Comment (Random-Liu) [inline pkg/kubelet/dockertools/fake_docker_client.go]: Same here.

File: pkg/kubelet/dockertools/fake_docker_client.go [hunk 4]
```diff
@@ -615,15 +634,9 @@ func (f *FakeDockerClient) PullImage(image string, auth dockertypes.AuthConfig,
...
-			ID:       image,
-			RepoTags: []string{image},
-			// Image size is required to be non-zero for CRI integration.
-			VirtualSize: fakeImageSize,
-			Size:        fakeImageSize,
-			Config:      &dockercontainer.Config{},
-		}
+		f.ImageInspects[image] = createImageInspectFromRef(image)
 		f.appendPulled(fmt.Sprintf("%s using %s", image, string(authJson)))
+		f.ImagesPulled = append(f.ImagesPulled, image)
```
Comment (Random-Liu) [inline pkg/kubelet/dockertools/fake_docker_client.go]: Can we also update `f.Images` here?

File: pkg/kubelet/dockertools/fake_docker_client.go [hunk 5]
```diff
@@ -713,44 +749,43 @@ func (f *FakeDockerClient) ResizeContainerTTY(id string, height, width int) erro
...
-// Pull records the image pull attempt, and optionally injects an error.
-func (f *FakeDockerPuller) Pull(image string, secrets []v1.Secret) (err error) {
-	f.Lock()
-	defer f.Unlock()
-	f.ImagesPulled = append(f.ImagesPulled, image)
-
-	if len(f.ErrorsToInject) > 0 {
-		err = f.ErrorsToInject[0]
-		f.ErrorsToInject = f.ErrorsToInject[1:]
+func createImageInspect(image dockertypes.Image) *dockertypes.ImageInspect {
```
Comment (Random-Liu) [inline pkg/kubelet/dockertools/fake_docker_client.go]: nit: createImageInspectFromImage

File: pkg/kubelet/dockertools/fake_docker_client.go [hunk 6]
```diff
@@ -713,44 +749,43 @@ func (f *FakeDockerClient) ResizeContainerTTY(id string, height, width int) erro
...
-	f.Lock()
-	defer f.Unlock()
-	if f.HasImages == nil {
-		return name, nil
-	}
-	for _, s := range f.HasImages {
-		if s == name {
-			return s, nil
-		}
+func createImage(inspect dockertypes.ImageInspect) *dockertypes.Image {
```
Comment (Random-Liu) [inline pkg/kubelet/dockertools/fake_docker_client.go]: nit: createImageFromImageInspect.

File: pkg/kubelet/dockertools/fake_docker_client.go [hunk 7]
```diff
@@ -33,6 +33,7 @@ import (
 	dockercontainer "github.com/docker/engine-api/types/container"
```
Comment (Random-Liu) [inline pkg/kubelet/dockertools/fake_docker_client.go]: Commit LGTM.

File: pkg/kubelet/dockertools/docker_manager_test.go [hunk 8]
```diff
@@ -1435,8 +1437,9 @@ func TestGetIPCMode(t *testing.T) {
 
 func TestSyncPodWithPullPolicy(t *testing.T) {
 	dm, fakeDocker := newTestDockerManagerWithRealImageManager()
-	puller := dm.dockerPuller.(*FakeDockerPuller)
-	puller.HasImages = []string{"foo/existing_one:v1", "foo/want:latest"}
+	fakeDocker.InjectImages([]dockertypes.Image{
+		{ID: "foo/existing_one:v1"}, {ID: "foo/want:latest"}})
```
Comment (Random-Liu) [inline pkg/kubelet/dockertools/docker_manager_test.go]: Unnecessary newline.

**Review by Random-Liu — COMMENTED** [review]
LGTM to me overall with some nits.

**Comment by Random-Liu** [comment]
Review status: 0 of 11 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/kubelet/dockertools/fake_docker_client.go, line 639 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42264#-KeBCJkk5BP4IKLEOHmi-r3-639:-KedvgnNJruyFaK9xix_:bm24jgg) ([raw file](https://github.com/kubernetes/kubernetes/blob/850f616a22246f2f2cc4f058b87e55d39a8aca58/pkg/kubelet/dockertools/fake_docker_client.go#L639)):*
<details><summary><i>Previously, Random-Liu (Lantao Liu) wrote…</i></summary><blockquote>

Can we also update `f.Images` here?
</blockquote></details>

It seems that this is not addressed. Is there any special consideration here?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/42264)*

**Comment by Random-Liu** [comment]
LGTM with one last nit.

**Comment by yujuhong** [comment]
> LGTM with one last nit.

Fixed.

**Comment by gmarek** [comment]
Thanks for doing this @yujuhong @Random-Liu

**Pull request merged**